drivers/i3c/i3c_dw.c

```diff
@@ -0,0 +1,1743 @@
+	sys_write32(INTR_ALL, config->regs + INTR_STATUS);
+
+	sys_write32(INTR_MASTER_MASK, config->regs + INTR_STATUS_EN);
+	sys_write32(INTR_MASTER_MASK, config->regs + INTR_SIGNAL_EN);
+}
+
+/**
+ * @brief DesignWare I3C init function
+ *
+ * @param dev: Pointer to the device structure
+ *
+ * @retval  0 if successful
+ * @retval -EINVAL if a structure/pointer is found NULL
+ *	    or if there exist LVR related error
+ * @retval -ENOTSUP if slaves no. exceeds more than supported
+ * @retval -ENOMEM if no such free address is found
+ * @retval -EBUSY if no I3C address slot is free
+ */
+static int32_t i3c_dw_init(const struct device *dev)
+{
+	const struct dw_i3c_config *config = DEV_CONFIG(dev);
+	struct dw_i3c_data *i3c = DEV_DATA(dev);
+	int32_t ret;
+
+	if (dev == NULL) {
+		LOG_ERR("%s Device ptr null !!\n", __func__);
+		return -EINVAL;
+	}
+
+	if (config->max_slv_devs > 4) {
+		LOG_ERR("%s Max slave dev count exceeded !\n", __func__);
+		return -ENOTSUP;
+	}
+
+	/* Allocate Memory */
+	i3c->base.addrs = (uint8_t *)k_malloc(config->max_slv_devs *
+			   sizeof(uint8_t));
+	i3c->slave = (struct dw_slave_info *)k_malloc(config->max_slv_devs
+		      * sizeof(struct dw_slave_info));
+	i3c->base.dev = dev;
+	i3c->base.ops = dev->api;
+
+	/* Initializing feasible address slots for DAA */
+	i3c_bus_init_addrslots(i3c->base.addrslots);
+
+	/* I3C pure, I2C pure or Mix */
+	ret = define_i3c_bus_mode(dev);
+	if (ret) {
+		return ret;
+	}
+
+	/* reset all */
+	sys_write32(RESET_CTRL_ALL, config->regs + RESET_CTRL);
+
+	/* get DAT, DCT pointer */
+	i3c->datstartaddr = DEVICE_ADDR_TABLE_ADDR(sys_read32(config->regs +
+			    DEVICE_ADDR_TABLE_POINTER));
+	i3c->dctstartaddr = DEVICE_CHAR_TABLE_ADDR(sys_read32(config->regs +
+			    DEV_CHAR_TABLE_POINTER));
+
+	/* free positions available among available slots */
+	i3c->free_pos = GENMASK(config->max_slv_devs - 1, 0);
+
+	/* Config timing registers */
+	dw_init_scl_timing(dev);
+
+	/* Self-assigning a dynamic address */
+	ret = master_addr_config(dev);
+	if (ret) {
+		return ret;
+	}
+
+	/* Enable Interrupt signals */
+	en_inter_sig_stat(dev);
+
+	/* disable IBI */
+	sys_write32(IBI_REQ_REJECT_ALL, config->regs + IBI_SIR_REQ_REJECT);
+	sys_write32(IBI_REQ_REJECT_ALL, config->regs + IBI_MR_REQ_REJECT);
+
+	/* disable hot-join */
+	sys_write32(sys_read32(DEV_CTRL(config)) | DEV_CTRL_HOT_JOIN_NACK,
+			DEV_CTRL(config));
+
+	/* enable I3C master */
+	sys_write32(sys_read32(DEV_CTRL(config)) | DEV_CTRL_ENABLE,
+			DEV_CTRL(config));
+
+	k_sem_init(&i3c->sem_xfer, 0, 1);
+	k_mutex_init(&i3c->mt);
+
+	return 0;
+}
+
+static const struct i3c_driver_api i3c_dw_api = {
+	.do_daa = dw_i3c_do_daa,
+	.i3c_xfers = dw_i3c_xfers,
+	.do_ccc = dw_i3c_send_ccc_cmd,
+	.ibi_enable = dw_i3c_enable_ibi,
+	.ibi_disable = dw_i3c_disable_ibi,
+};
+
+#define I3C_DW_IRQ_HANDLER(n)				\
+	static void i3c_dw_irq_config_##n(void)		\
+{							\
+	IRQ_CONNECT(DT_INST_IRQN(n),			\
+			1,				\
```

Comment:
just fixed priority? this should be defined in the dts

Reply:
Okay.